File: Restore-IP-Addresses/RIA.py
testStr = "25525511135"
testStr2 = "0000"
testStr3 = "010010"
# answer 3: ["0.10.0.10","0.100.1.0"]
testStr4 = "101023"
# answer 4: ["1.0.10.23","1.0.102.3","10.1.0.23","10.10.2.3","101.0.2.3"]


class Solution(object):

    def getNext(self, sLeft):
        tempStore = []
        for i in range(3):
            try:
                # if int(sLeft[i + 1]) < 3:
                tempStore.append(sLeft[0:i + 1])
            except:
                continue
        return [[ss] + [sLeft[len(ss):]] for ss in tempStore]

    def parseList(self, l):
        reStr = l[0]
        for i in l[1:]:
            reStr += "." + i
        return reStr

    def restoreIpAddresses(self, s):
        """
        :type s: str
        :rtype: List[str]
        """
        listPool = [[s]]
        for i in range(3):
            tempList = []
            for l in listPool:
                for ll in self.getNext(l[-1]):
                    tempList.append(
                        l[:-1] + ll)
            listPool = tempList

        reList = []
        for l in listPool:
            if l[-1] == "":
                continue
            thisBool = True
            for ii in l:
                if int(ii) > 255 or \
                   len(ii) > 3 or \
                   (len(ii) > 1 and ii[0] == "0"):
                    thisBool = False
            if thisBool is True:
                # print(l)
                reList.append(self.parseList(l))

        return reList


if __name__ == "__main__":
    a = Solution()
    a.restoreIpAddresses(testStr)

File: Restore-IP-Addresses/RIA.swift
let testStr = "25525511135"
let testStr2 = "0000"
let testStr3 = "010010"
let testStr4 = "101023"

class Solution {
    func cutIp(_ s:String) -> [[String]] {
        let lenStr:Int = s.characters.count
        let iterTime:Int = min(lenStr, 4)
        var reStrList = [[String]]()

        for i in 1...(iterTime - 1) {
            let startIndex = s.startIndex
            let endIndex = s.endIndex
            let cutIndex = s.index(s.startIndex, offsetBy:i)

            reStrList.append(
              [s[startIndex..<cutIndex],s[cutIndex..<endIndex]]
            )
        }
        //print(reStrList)
        return reStrList
    }

    func nextCut(_ strL:[String]) -> [[String]] {
        if strL.count == 1{
            return cutIp(strL.last!)
        }

        var reStrList = [[String]]()
        if strL.last!.characters.count == 1{
            reStrList.append(strL)
        }else {
            for i in cutIp(strL.last!) {
                var temp = strL.dropLast()
                temp.append(contentsOf:i) // temp is a slice
                reStrList.append(
                  Array(temp)
                )
            }
        }

        return reStrList
    }

    func mergeIp(_ l:[String]) -> String{
        var str = l.reduce("", {$0 + "." + $1})
        str.remove(at:str.startIndex)
        return str
    }

    func headZero (_ sL:[String]) -> Bool{
        for s in sL {
            if (s.characters[s.startIndex] == "0"
                  && s.characters.count > 1) {
                return false
            }
        }
        return true
    }
    
    func less255(_ sL:[String]) -> Bool{
        for s in sL {
            if Int(s)! > 255{
                return false
            }
        }
        return true
    }
    
    
    func restoreIpAddresses(_ s:String) -> [String] {
        var tempStr = [[s]]

        for _ in 1...3 {
            var temp = [[String]]()
            
            for l in tempStr {
                temp.append(contentsOf: self.nextCut(l))
            }
            tempStr = temp
        }

        var reStr:[String] = []
        for l in tempStr {
            if (l.count == 4 &&
                  self.less255(l) &&
                  self.headZero(l)) {
                reStr.append(self.mergeIp(l))
            }
        }
        
        //print(tempStr)
        return reStr
    }
}


var a = Solution()

print(a.restoreIpAddresses(testStr))
print(a.restoreIpAddresses(testStr2))
print(a.restoreIpAddresses(testStr3))
print(a.restoreIpAddresses(testStr4))
